### PR TITLE
applications: nrf5340_audio: dfu clean up

### PR DESCRIPTION
**Pull request by pirun**
Fix assertation when DFU server disconnect with CIS Headset config.
Remove "TODO:" in overlay files.
Remove kconfig kconfig assigned warning.
Hardreset while programming external flash layout.
Remove shebang line in python scripts.
Remove unused DEFINES in buildprog.py.
Remove generated binary files in /bin, put them back to build/zephyr.
Use empty_net as b0n placeholder, not building hci_rpmsg anymore.
Remove DFU dependence of buildprog.py, ble5-ctr-rpmsg-sign now work
with west build system.
Fix BIS headset could not scan gateway occasionally.

Signed-off-by: Pirun Lee <pirun.lee@nordicsemi.no>

### CODE REVIEW

File: boards/arm/nrf5340_audio_dk_nrf5340/nrf5340_audio_dk_nrf5340_cpuapp_defconfig [hunk 1]
```diff
@@ -4,74 +4,19 @@ CONFIG_SOC_SERIES_NRF53X=y
 CONFIG_SOC_NRF5340_CPUAPP_QKAA=y
 CONFIG_BOARD_NRF5340_AUDIO_DK_NRF5340_CPUAPP=y
 
-CONFIG_BOARD_ENABLE_DCDC_APP=y
```
Comment (koffes) [inline boards/arm/nrf5340_audio_dk_nrf5340/nrf5340_audio_dk_nrf5340_cpuapp_defconfig:7]: Does the boardfiles need to change in this PR? 
If so, please see if you can use the changes from https://github.com/zephyrproject-rtos/zephyr/pull/49720.
This needs to be done with caution as we have large boardfile changes happening in two places.

**Review by koffes — CHANGES_REQUESTED** [review]

File: applications/nrf5340_audio/README.rst [hunk 2]
```diff
@@ -870,19 +870,12 @@ FOTA build files
...
 
-* For multi-image DFU, the file name refers to ``dfu_application`` and includes *<Role>*, *<Application version>*, and *<Revision>* in the file name (``dev_<Role>_build_<Application version>_dfu_application_<Revision>.zip``, for example :file:`dev_headset_build_debug_dfu_application_0.0.0+0.zip`).
+* For multi-image DFU, the file name refers to ``dfu_application.zip``.
   This file updates two cores with one single file.
-* For single-image DFU, the bin file for the application core refers in the name to ``app_update`` and includes *<Role>*, *<Application version>*, and *<Revision>* in the file name (``dev_<Role>_build_<Application version>_app_update_<Revision>.bin``, for example :file:`dev_headset_build_debug_app_update_0.0.0+0.bin`).
-  The bin file for the network core refers in the name to ``net_core_app_update`` and includes *<FW_VERSION>* instead of *<Revision>* in the file name (``dev_<Role>_build_<Application version>_net_core_app_update_<FW_VERSION>.bin``, for example :file:`dev_headset_build_debug_net_core_app_update_1.bin`).
+* For single-image DFU, the bin file for the application core refers in the name to ``app_update.bin``.
```
Comment (koffes) [inline applications/nrf5340_audio/README.rst:875]: Should we even mention single image DFU?

File: applications/nrf5340_audio/tools/buildprog/buildprog.py [hunk 3]
```diff
@@ -45,8 +44,6 @@
 TARGET_RELEASE_FOLDER = "build_release"
 TARGET_DEBUG_FOLDER = "build_debug"
 
-PRISTINE_FLAG = " --pristine"
```
Comment (koffes) [inline applications/nrf5340_audio/tools/buildprog/buildprog.py:48]: Why is this removed?

File: boards/arm/nrf5340_audio_dk_nrf5340/nrf5340_audio_dk_nrf5340_cpuapp_defconfig [hunk 4]
```diff
@@ -4,74 +4,19 @@ CONFIG_SOC_SERIES_NRF53X=y
...
 
-# Bluetooth Low Energy
-CONFIG_BT=y
+# enable console
+CONFIG_CONSOLE=y
+CONFIG_UART_CONSOLE=y
 
-# Add drivers for Cirrus Logic HW Codec
-CONFIG_HW_CODEC_CIRRUS_LOGIC=y
+CONFIG_PINCTRL=y
```
Comment (koffes) [inline boards/arm/nrf5340_audio_dk_nrf5340/nrf5340_audio_dk_nrf5340_cpuapp_defconfig:22]: Missing newline

**Review by koffes — CHANGES_REQUESTED** [review]
I need some more time to go through this PR in detail.

File: boards/arm/nrf5340_audio_dk_nrf5340/nrf5340_audio_dk_nrf5340_cpuapp_defconfig (hunk 1 above)
Comment (koffes) [reply boards/arm/nrf5340_audio_dk_nrf5340/nrf5340_audio_dk_nrf5340_cpuapp_defconfig:7]: Another option is to bring the upstream boardfiles into NCS, temporarily.

File: boards/arm/nrf5340_audio_dk_nrf5340/nrf5340_audio_dk_nrf5340_cpunet_defconfig [hunk 5]
```diff
@@ -16,11 +16,9 @@ CONFIG_PINCTRL=y
...
-# HCI_RPMSG
-CONFIG_RPMSG_SERVICE=y
-CONFIG_RPMSG_SERVICE_MODE_REMOTE=y
+# Enable uart driver
+CONFIG_SERIAL=y
 
-CONFIG_HEAP_MEM_POOL_SIZE=8192
-
-CONFIG_BT=y
+# enable console
```
Comment (rick1082) [inline boards/arm/nrf5340_audio_dk_nrf5340/nrf5340_audio_dk_nrf5340_cpunet_defconfig:22]: Capital "E" for enable.
Since we don't expose the UART log from controller to public, would this part of the setting only for B0n?

File: boards/arm/nrf5340_audio_dk_nrf5340/nrf5340_audio_dk_nrf5340_cpuapp_defconfig [hunk 6]
```diff
@@ -4,74 +4,19 @@ CONFIG_SOC_SERIES_NRF53X=y
...
-# exFAT enabled to support longer file names and higher transfer speed
-CONFIG_FS_FATFS_EXFAT=y
 
-# Set the maximum file name length to 255
-CONFIG_FS_FATFS_MAX_LFN=255
+CONFIG_SERIAL=y
 
-# Bluetooth Low Energy
-CONFIG_BT=y
+# enable console
```
Comment (rick1082) [inline boards/arm/nrf5340_audio_dk_nrf5340/nrf5340_audio_dk_nrf5340_cpuapp_defconfig:18]: Capitial "E" for enable

**Review by rick1082 — CHANGES_REQUESTED** [review]

File: applications/nrf5340_audio/README.rst (hunk 2 above)
Comment (pirun) [reply applications/nrf5340_audio/README.rst:875]: > Should we even mention single image DFU?

I think so, we still have internal layout option.

File: applications/nrf5340_audio/tools/buildprog/buildprog.py (hunk 3 above)
Comment (pirun) [reply applications/nrf5340_audio/tools/buildprog/buildprog.py:48]: > Why is this removed?

Because we don't use this DEFINE in buildprog.py.
When --pristine is assigned to buildprog.py, we use two methods to make a clean build.

1. remove build tree by ``shutil.rmtree(dest_folder)``.
2. ``west build -p`` this is the same thing ``west build --pristine``

File: boards/arm/nrf5340_audio_dk_nrf5340/nrf5340_audio_dk_nrf5340_cpuapp_defconfig (hunk 1 above)
Comment (pirun) [reply boards/arm/nrf5340_audio_dk_nrf5340/nrf5340_audio_dk_nrf5340_cpuapp_defconfig:7]: Yes , to remove those "=n" kconfig in child image(mcuboot) overlay, the boardfiles need to be changed.
As the comment in https://github.com/zephyrproject-rtos/zephyr/pull/49720.
> The basic default for a board needs to have all peripherals disabled and only allows for UART and timer to be enabled

What I do is to copy your changes of ``nrf5340_audio_dk_nrf5340_cpuapp_defconfig`` and ``nrf5340_audio_dk_nrf5340_cpunet_defconfig`` from upstream PR except license declaration.

File: boards/arm/nrf5340_audio_dk_nrf5340/nrf5340_audio_dk_nrf5340_cpuapp_defconfig (hunk 6 above)
Comment (pirun) [reply boards/arm/nrf5340_audio_dk_nrf5340/nrf5340_audio_dk_nrf5340_cpuapp_defconfig:18]: As @koffes mention.

> This needs to be done with caution as we have large boardfile changes happening in two places.

I copied defconfig straight from https://github.com/zephyrproject-rtos/zephyr/pull/49720.

File: boards/arm/nrf5340_audio_dk_nrf5340/nrf5340_audio_dk_nrf5340_cpunet_defconfig (hunk 5 above)
Comment (pirun) [reply boards/arm/nrf5340_audio_dk_nrf5340/nrf5340_audio_dk_nrf5340_cpunet_defconfig:22]: Yes, console only affect when b0n print boot banner.

File: boards/arm/nrf5340_audio_dk_nrf5340/nrf5340_audio_dk_nrf5340_cpuapp_defconfig (hunk 4 above)
Comment (pirun) [reply boards/arm/nrf5340_audio_dk_nrf5340/nrf5340_audio_dk_nrf5340_cpuapp_defconfig]: Added.

File: boards/arm/nrf5340_audio_dk_nrf5340/nrf5340_audio_dk_nrf5340_cpuapp_defconfig (hunk 1 above)
Comment (koffes) [reply boards/arm/nrf5340_audio_dk_nrf5340/nrf5340_audio_dk_nrf5340_cpuapp_defconfig:7]: Great!

File: applications/nrf5340_audio/README.rst (hunk 2 above)
Comment (koffes) [reply applications/nrf5340_audio/README.rst:875]: Could we then have a disclaimer here as well? Or link to the documentation? We really do not want to use this approach.

File: applications/nrf5340_audio/README.rst [hunk 7]
```diff
@@ -1473,4 +1466,4 @@ Legal notices for the nRF5340 Audio DK
 
 .. |net_core_hex_note| replace:: The network core for both gateway and headsets is programmed with the precompiled Bluetooth Low Energy Controller binary file :file:`ble5-ctr-rpmsg_<XYZ>.hex`, where *<XYZ>* corresponds to the controller version, for example :file:`ble5-ctr-rpmsg_3216.hex`.
    This file includes the LE Audio Controller Subsystem for nRF53 and is provided in the :file:`applications/nrf5340_audio/bin` directory.
-   If :ref:`DFU <nrf53_audio_app_configuration_configure_fota>` is enabled, the subsystem's binary file will include variables from the naming patterns for FOTA build files (:file:`dev_<Role>_build_<Application version>_ble5-ctr_CPUNET.hex`).
+   If :ref:`DFU <nrf53_audio_app_configuration_configure_fota>` is enabled, the subsystem's binary file will be generate in :file:`build/zephyr/` folder (:file:`net_core_app_signed.hex`).
```
Comment (greg-fer) [inline applications/nrf5340_audio/README.rst:1469]:
```suggestion
   If :ref:`DFU is enabled <nrf53_audio_app_configuration_configure_fota>`, the subsystem's binary file will be generated in the :file:`build/zephyr/` directory and will be called :file:`net_core_app_signed.hex`.
```

File: applications/nrf5340_audio/README.rst [hunk 8]
```diff
@@ -870,19 +870,12 @@ FOTA build files
 
 The generated FOTA build files use the following naming patterns:
 
-* For multi-image DFU, the file name refers to ``dfu_application`` and includes *<Role>*, *<Application version>*, and *<Revision>* in the file name (``dev_<Role>_build_<Application version>_dfu_application_<Revision>.zip``, for example :file:`dev_headset_build_debug_dfu_application_0.0.0+0.zip`).
+* For multi-image DFU, the file name refers to ``dfu_application.zip``.
   This file updates two cores with one single file.
```
Comment (greg-fer) [inline applications/nrf5340_audio/README.rst:874]:
```suggestion
* For multi-image DFU, the file is called ``dfu_application.zip``.
  This file updates two cores with one single file.
```

File: applications/nrf5340_audio/README.rst (hunk 2 above)
Comment (greg-fer) [inline applications/nrf5340_audio/README.rst:875]:
```suggestion
* For single-image DFU, the bin file for the application core is called ``app_update.bin``.
```

File: applications/nrf5340_audio/README.rst [hunk 9]
```diff
@@ -870,19 +870,12 @@ FOTA build files
...
-* For single-image DFU, the bin file for the application core refers in the name to ``app_update`` and includes *<Role>*, *<Application version>*, and *<Revision>* in the file name (``dev_<Role>_build_<Application version>_app_update_<Revision>.bin``, for example :file:`dev_headset_build_debug_app_update_0.0.0+0.bin`).
-  The bin file for the network core refers in the name to ``net_core_app_update`` and includes *<FW_VERSION>* instead of *<Revision>* in the file name (``dev_<Role>_build_<Application version>_net_core_app_update_<FW_VERSION>.bin``, for example :file:`dev_headset_build_debug_net_core_app_update_1.bin`).
+* For single-image DFU, the bin file for the application core refers in the name to ``app_update.bin``.
+  The bin file for the network core refers in the name to ``net_core_app_update.bin``.
```
Comment (greg-fer) [inline applications/nrf5340_audio/README.rst:876]:
```suggestion
  The bin file for the network core is called ``net_core_app_update.bin``.
```

File: applications/nrf5340_audio/README.rst (hunk 2 above)
Comment (pirun) [reply applications/nrf5340_audio/README.rst:875]: We have already stated that in caution part.
https://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-9275/nrf/applications/nrf5340_audio/README.html#configuring-fota-upgrades

File: applications/nrf5340_audio/README.rst [hunk 10]
```diff
@@ -870,19 +870,12 @@ FOTA build files
...
   In this scenario, the cores are updated one by one with two separate files in two actions.
```
Comment (greg-fer) [inline applications/nrf5340_audio/README.rst:878]: Add one extra line after this list:
```suggestion
  In this scenario, the cores are updated one by one with two separate files in two actions.

See :ref:`app_build_output_files` for more information about the image files.

```

File: applications/nrf5340_audio/README.rst (hunk 7 above)
Comment (greg-fer) [reply applications/nrf5340_audio/README.rst:1469]: @tejlmand , net_core_app_signed.hex is not listed as one of the output files here: https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/app_build_system.html#output-build-files 

Should we add it to one of the tables?

**Review by greg-fer — CHANGES_REQUESTED** [review]

File: applications/nrf5340_audio/dfu/conf/Kconfig.dfu [hunk 11]
```diff
@@ -41,7 +32,8 @@ config BOOTLOADER_MCUBOOT
 	bool
 	default y
 
-config NCS_INCLUDE_RPMSG_CHILD_IMAGE
+
+config EMPTY_NET_CORE_CHILD_IMAGE
```
Comment (koffes) [inline applications/nrf5340_audio/dfu/conf/Kconfig.dfu:36]: Could you elaborate on this?

File: applications/nrf5340_audio/dfu/empty_net/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,12 @@
+#
```
Comment (koffes) [inline applications/nrf5340_audio/dfu/empty_net/CMakeLists.txt:1]: Why is the empty NET required?

File: applications/nrf5340_audio/dfu/empty_net/src/main.c [hunk 13]
```diff
@@ -0,0 +1,12 @@
...
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/kernel.h>
+
+void main(void)
+{
+	;
```
Comment (koffes) [inline applications/nrf5340_audio/dfu/empty_net/src/main.c:11]: I believe the NET core should be put in sleep if not used, or is this entire file a dummy?

File: applications/nrf5340_audio/src/modules/Kconfig.defaults [hunk 14]
```diff
@@ -45,6 +45,15 @@ config USB_DEVICE_MANUFACTURER
 	string
 	default "Nordic Semiconductor AS"
 
+## Avoid redundant warnings for endpoint setting in USB stack
```
Comment (koffes) [inline applications/nrf5340_audio/src/modules/Kconfig.defaults:48]: Do these need to be part of this PR?

**Review by koffes — CHANGES_REQUESTED** [review]

File: applications/nrf5340_audio/dfu/conf/Kconfig.dfu (hunk 11 above)
Comment (pirun) [reply applications/nrf5340_audio/dfu/conf/Kconfig.dfu:36]: This is dummy netcore appliation for generating b0n.
We used hci_rpmsg as dummy app before.
Since we use prebuilt controller, it makes no sense to build hci_rpmsg.
So I decide to use truly dummy one.

File: applications/nrf5340_audio/dfu/empty_net/CMakeLists.txt (hunk 12 above)
Comment (pirun) [reply applications/nrf5340_audio/dfu/empty_net/CMakeLists.txt:1]: For generating NET bootloader and signature.

File: applications/nrf5340_audio/dfu/empty_net/src/main.c (hunk 13 above)
Comment (pirun) [reply applications/nrf5340_audio/dfu/empty_net/src/main.c:11]: This is dummy application. We don't program it.

File: applications/nrf5340_audio/src/modules/Kconfig.defaults (hunk 14 above)
Comment (pirun) [reply applications/nrf5340_audio/src/modules/Kconfig.defaults:48]: One of major purposes of this PR is to elimate kconfig warning messages while compiling.
These 2 kconfigs are only applicable when device is gateway.
They will generate kconfig warnings when compiling headset since headset doesn't enable USB driver.

**Review by rick1082 — APPROVED** [review]

**Review by greg-fer — APPROVED** [review]

**Comment by rlubos** [comment]
@koffes Please revisit.

File: applications/nrf5340_audio/dfu/conf/Kconfig.dfu (hunk 11 above)
Comment (koffes) [reply applications/nrf5340_audio/dfu/conf/Kconfig.dfu:36]: Thanks. Can we rename or add a comment that this is a dummy?

File: applications/nrf5340_audio/dfu/empty_net/src/main.c (hunk 13 above)
Comment (koffes) [reply applications/nrf5340_audio/dfu/empty_net/src/main.c:11]: Add a comment in main. Suggest:
```suggestion
	; /* Dummy application. Unused main */
```

**Review by koffes — COMMENTED** [review]
Nice work. I have two outstanding comments.

File: applications/nrf5340_audio/dfu/conf/Kconfig.dfu (hunk 11 above)
Comment (pirun) [reply applications/nrf5340_audio/dfu/conf/Kconfig.dfu:35]: Add

> 	bool "Dummy Net core application"

**Review by koffes — APPROVED** [review]
